src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -170,42 +171,78 @@ private JvmObjectReference CreateBroadcast(SparkContext sc, T value)
             bool encryptionEnabled = bool.Parse(
                 sc.GetConf().Get("spark.io.encryption.enabled", "false"));
 
+            var _pythonBroadcast = (JvmObjectReference)javaSparkContext.Jvm.CallStaticJavaMethod(
+                "org.apache.spark.api.python.PythonRDD",
+                "setupBroadcast",
+                _path);
+
             if (encryptionEnabled)
             {
-                throw new NotImplementedException("Broadcast encryption is not supported yet.");
+                var pair = (JvmObjectReference[])_pythonBroadcast.Invoke("setupEncryptionServer");
+
+                using (ISocketWrapper socket = SocketFactory.CreateSocket())
+                {
+                    socket.Connect(
+                        IPAddress.Loopback,
+                        (int)pair[0].Invoke("intValue"), // port number
+                        (string)pair[1].Invoke("toString")); // secret
+                    WriteToStream(value, socket.OutputStream, true);
+                    socket.OutputStream.Flush();
+                }
+                _pythonBroadcast.Invoke("waitTillDataReceived");
             }
             else
             {
-                WriteToFile(value);
+                WriteToStream(value, null, false);
             }
 
-            var pythonBroadcast = (JvmObjectReference)javaSparkContext.Jvm.CallStaticJavaMethod(
-                "org.apache.spark.api.python.PythonRDD",
-                "setupBroadcast",
-                _path);
-
-            return (JvmObjectReference)javaSparkContext.Invoke("broadcast", pythonBroadcast);
+            return (JvmObjectReference)javaSparkContext.Invoke("broadcast", _pythonBroadcast);
         }
 
+        /// TODO: This is not performant in the case of Broadcast encryption as it writes to stream
+        /// only after serializing the whole value, instead of serializing and writing in chunks
+        /// like Python.
         /// <summary>
-        /// Function that creates a file in _path to store the broadcast value in the given path.
+        /// Function to write the broadcast value into the stream.
         /// </summary>
-        /// <param name="value">Broadcast value to be written to the file</param>
-        private void WriteToFile(object value)
+        /// <param name="value">Broadcast value to be written to the stream</param>
+        /// <param name="stream">Stream to write value to</param>
+        /// <param name="isEncrypted">Boolean value to check if broadcast encrytion is set</param>
+        private void WriteToStream(object value, Stream stream, bool isEncrypted)
         {
-            using FileStream f = File.Create(_path);
-            Dump(value, f);
+            if (isEncrypted)
+            {
+                Dump(value, stream, isEncrypted);
```

Comment:
Hmm. I didn't mean it this way...

Can you revert back the changes around here? I lost track of what happened.

As far as I remember, you shouldn't change the `Dump` but call it from `WriteToStream`, and keep `WriteToFile`.

Reply:
Reverted changes. I thought you wanted to call `Dump` from both `WriteToStream` and `WriteToFile` which required some changes to be made to `Dump`, since encryption needs some massaging of data before writing to stream.

Reply:
No, I meant just calling `Dump` without modifying it since serializing using the formatter is the same.

Reply:
Oh sorry, I thought you meant passing the socket stream, got it, changed, thanks!